File: Python/Hard/3600 Maximize Spanning Tree Stability with Upgrades.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rx, ry = self.find(x), self.find(y)
        if rx == ry: 
            return False
        if self.rank[rx] > self.rank[ry]:
            self.parent[ry] = rx
        elif self.rank[rx] < self.rank[ry]:
            self.parent[rx] = ry
        else:
            self.parent[ry] = rx
            self.rank[rx] += 1
        return True

class Solution:
    def maxStability(self, n: int, edges: list[list[int]], k: int) -> int:
        uf = UnionFind(n)
        cnt = 0
        min_s = float('inf')
        
        # Pass 1: Add all mandatory edges
        for u, v, s, m in edges:
            if m == 0: 
                continue
            if not uf.union(u, v): 
                return -1 # Cycle detected in mandatory edges
            cnt += 1
            min_s = min(min_s, s)
            
        # Pass 2: Sort optional edges descending
        edges.sort(key=lambda x: x[2], reverse=True)
        
        for u, v, s, m in edges:
            if m == 1: 
                continue
            if not uf.union(u, v): 
                continue
            cnt += 1
            
            # Tracking the bottleneck of un-upgraded vs upgraded edges
            if cnt == (n - 1) - k:
                min_s = min(min_s, s)
            elif cnt == n - 1:
                min_s = min(min_s, 2 * s)
                
        return min_s if cnt == n - 1 else -1

File: Java/Hard/3600 Maximize Spanning Tree Stability with Upgrades.java
import java.util.Arrays;

class Solution {
    static class UnionFind {
        int[] parent;
        int[] rank;

        public UnionFind(int n) {
            parent = new int[n];
            rank = new int[n];
            for (int i = 0; i < n; i++) parent[i] = i;
        }

        public int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        public boolean union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX == rootY) return false;
            
            if (rank[rootX] > rank[rootY]) {
                parent[rootY] = rootX;
            } else if (rank[rootX] < rank[rootY]) {
                parent[rootX] = rootY;
            } else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
            return true;
        }
    }

    public int maxStability(int n, int[][] edges, int k) {
        UnionFind uf = new UnionFind(n);
        int cnt = 0;
        int minStrength = Integer.MAX_VALUE;

        // Pass 1: Add all mandatory edges
        for (int[] edge : edges) {
            if (edge[3] == 0) continue;
            if (!uf.union(edge[0], edge[1])) return -1;
            cnt++;
            minStrength = Math.min(minStrength, edge[2]);
        }

        // Sort remaining optional edges in descending order by strength
        Arrays.sort(edges, (a, b) -> Integer.compare(b[2], a[2]));

        // Pass 2: Fill in the remaining needed components 
        for (int[] edge : edges) {
            if (edge[3] == 1) continue;
            if (!uf.union(edge[0], edge[1])) continue;
            cnt++;
            
            if (cnt == (n - 1) - k) {
                minStrength = Math.min(minStrength, edge[2]);
            } else if (cnt == n - 1) {
                minStrength = Math.min(minStrength, 2 * edge[2]);
            }
        }

        return (cnt == n - 1) ? minStrength : -1;
    }
}
